woocommerce/woocommerce
# [Experimental] Surface additional checkout fields on Order Confirmation page

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Note that this PR is based on this branch:
https://github.com/woocommerce/woocommerce/pull/43274

Implements a new _Additional Fields Wrapper_ block that contains a customisable heading, and an _Additional Fields_ block. This lists any custom checkout fields added to an order via block based checkout from the `contact` and `additional` areas.

In addition, this updates the shipping and billing address blocks to render custom fields added within the `address` area.

![Screenshot 2024-01-10 at 12 39 07](https://github.com/woocommerce/woocommerce/assets/90977/e96df931-29b8-48ef-84de-ae47312d787b)

@elizaan36 would you mind reviewing the appearance of these changes? The original issue lacked a design so I took cues from the other blocks and made it align with the columns of shipping/billing address. Note I also removed the italics in the address to match the screenshots in https://github.com/woocommerce/woocommerce/issues/42776 and removed the placeholder skeleton in favour of text for additional information:

![Screenshot 2024-01-10 at 12 41 16](https://github.com/woocommerce/woocommerce/assets/90977/9b7c90b0-317d-4814-9411-33ba66e477b7)

(this is the editor view)

Closes #42776



### How to test the changes in this Pull Request:

Before testing, you'll need to register some custom checkout fields. You can use the following snippet.

```php
add_action(
	'woocommerce_loaded',
	function() {
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/some-new-billing-field',
				'label'    => __( 'Custom address field', 'woocommerce' ),
				'location' => 'address',
				'type'     => 'text',
			),
		);

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/some-new-contact-field',
				'label'    => __( 'Custom new contact field', 'woocommerce' ),
				'location' => 'contact',
				'type'     => 'text',
			),
		);

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-on-porch',
				'label'    => __( 'Please leave my package on the porch if I\'m not home', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'checkbox',
			),
		);

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/location-on-porch',
				'label'    => __( 'Describe where we should hide the parcel', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'text',
			)
		);
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/leave-with-neighbor',
				'label'    => __( 'Which neighbor should we leave it with if unable to hide?', 'woocommerce' ),
				'location' => 'additional',
				'type'     => 'select',
				'options'  => array(
					array(
						'label' => 'Neighbor to the left',
						'value' => 'left',
					),
					array(
						'label' => 'Neighbor to the right',
						'value' => 'right',
					),
					array(
						'label' => 'Neighbor across the road',
						'value' => 'across',
					),
					array(
						'label' => 'Do not leave with a neighbor',
						'value' => 'none',
					),
				),
			)
		);
	}
);
```

1. Use the above snippet.
2. Edit the Order Confirmation template in site editor. Revert customisations. This will cause _Additional Checkout Fields_ blocks to be inserted.
3. Add something to your cart and go through block checkout, populating the above new fields when doing so.
4. When the confirmation shows, ensure the custom fields are shown below the addresses.
5. Edit the template in site editor again. This time change the heading and/or move the blocks around. Save changes.
6. Checkout again. Ensure the changes made to the template persist on the frontend.
7. Remove the snippet you added previously.
8. Edit the template. Additional Checkout Fields blocks and the heading are hidden.
9. Checkout again. Additional Checkout Fields blocks and the heading are hidden.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

This functionality is behind a feature flag.

</details>
